quickwit-search/src/retry/mod.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+pub mod search;
+pub mod search_stream;
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use quickwit_proto::{FetchDocsRequest, LeafSearchRequest, LeafSearchStreamRequest};
+
+use crate::client_pool::Job;
+use crate::{ClientPool, SearchClientPool, SearchServiceClient};
+
+/// A retry policy to evaluate if a request should be retried.
+/// A retry can be made either on an error or on a partial success.
+pub trait RetryPolicy<Request, Response, Error>: Sized
+where Request: RequestOnSplit + Clone
+{
+    /// Returns a retry request in case of retry.
+    fn retry_request(&self, req: &Request, _result: Result<&Response, &Error>) -> Option<Request>;
+}
+
+/// Default retry policy:
+/// - All responses are treated as success.
+/// - All errors are retryable and the retry request is the same as the original one.
+pub struct DefaultRetryPolicy {}
+
+impl<Request, Response, Error> RetryPolicy<Request, Response, Error> for DefaultRetryPolicy
+where Request: RequestOnSplit + Clone
+{
+    fn retry_request(&self, req: &Request, result: Result<&Response, &Error>) -> Option<Request> {
+        match result {
+            Ok(_) => None,
+            Err(_) => Some(req.clone()),
+        }
+    }
+}
+
+// Select a new client from the client pool by the following oversimplified policy:
+// 1. Take the first split_id of the request
+// 2. Ask for a relevant client for that split while excluding the failing client.
+pub async fn retry_client<Request>(
+    client_pool: &Arc<SearchClientPool>,
+    failing_client: &SearchServiceClient,
```

Comment:
shouldn't this be a list of failing clients to exclude all failing clients as alternative (except all are failing)

Reply:
it would be better yes, it's clearly oversimplified. The main advantage of this PR is: we do a very simple retry for every type of request while keeping the retry code limited. But it's a babystep clearly.

The plan is to improve that point in another PR.

But I need to think more deeply about this. I want also to think about how we want to execute requests. For example, we may want to retry a request if we think that a node is taking too much time. We would need some sort of tasks executor... I need to read some code from bigger projects to know what we should do there.